packages/flutter/lib/src/rendering/sliver_list.dart

```diff
@@ -90,8 +90,20 @@ class RenderSliverList extends RenderSliverMultiBoxAdaptor {
     // it's possible for a child to get removed without notice.
     RenderBox leadingChildWithLayout, trailingChildWithLayout;
 
-    // Find the last child that is at or before the scrollOffset.
     RenderBox earliestUsefulChild = firstChild;
+    // We rely on firstChild to have accurate layout offset. Otherwise, we
+    // have to recalculate layout offset from the start.
+    if (childScrollOffset(firstChild) == null) {
+      while (earliestUsefulChild != null) {
+        assert(firstChild == earliestUsefulChild);
+        earliestUsefulChild = insertAndLayoutLeadingChild(childConstraints, parentUsesSize: true);
```

Comment:
Re-layouting the entire list seems like a lot of work and the list may be very long.

In the performRebuild of the SliverMultiBoxAdapterElement when we move a keyed element from one slot to another, could we just remember its layout offset for the slot the element was moved from and then re-use that layout offset for the new element we inflate in that slot?

https://github.com/flutter/flutter/blob/357d02c87b75d74382db212812e85a5d7d93dd7f/packages/flutter/lib/src/widgets/sliver.dart#L1162

Reply:
This might create a different behavior when prepend a new item.
Think about this case


```
scroll offset is 1000 and assume all item has 100 paintExtent
-----viewport-----
item10 layoutoffset 1000 paintExtent 100
item11  layoutoffset 1100 paintExtent 100
item12  layoutoffset 1200 paintExtent 100
-------------------


Later on we prepend a bigger item with paintExtent of 500.
If we always inherit layoutoffset of slot, every element will be shifted by 100 pixels
-----viewport-----
item9 layoutoffset 1000 paintExtent 100
item10  layoutoffset 1100 paintExtent 100
item11  layoutoffset 1200 paintExtent 100
-------------------

If we set it to null and recalculate, every element should be shifted by 500 pixels
-----viewport-----
item6 layoutoffset 1000 paintExtent 100
item7  layoutoffset 1100 paintExtent 100
item8  layoutoffset 1200 paintExtent 100
-------------------
```
This difference will become more obvious when the item are more diverse.

Reply:
That would get corrected as you scroll up, though, right? Just like it currently does if you use unkeyed children?

Reply:
This might not be to big of a problem. On the other hand, I am also wondering how often this will happen

Reply:
Yes, it will be corrected

Reply:
I think the current behavior is acceptable, it may even be the desired behavior since the promise is that the SliverList is lazy and as long as something is not on screen we don't know how big it is per design.

Reply:
fixed, but I still have to keep this logic here since there is still some cases we won't know the layout offset.

Reply:
What case still requires this code path?

Reply:
The test 'SliverList should recalculate inaccurate layout offset case 2' will requires this code path.
If you have
```
-----viewport-----
item10 layoutoffset 1000
item11  layoutoffset 1100
item12  layoutoffset 1200
-------------------
```
and you reorder item 12 to item3 which is not in the viewport. It will not know the current layout offset.

```
-----viewport-----
item12(index 3) layoutoffset ?
item10 layoutoffset 1000
item11  layoutoffset 1100
-------------------

Reply:
This is unfortunate. I wish we had a way around walking the full list. Especially since (as shown by your test) all this work is essentially thrown away anyways. 

I wonder if we could collect the renderobjects/elements that get moved to indexes before the current start index as garbage so they get removed from the tree for now and we are guaranteed to start with a child that does have a valid layoutOffset.

Reply:
There definitely will be case I have to rewalk the tree. However, there are one thing i can do to avoid doing that as much as possible.

try to locate the first child with (layout offset != null && layout offset <= scroll offset + cacheExtent), and use it to find the early useful child, and do a early garbage collect right after that. I think that is a good idea. I will implement that

Reply:
What are the cases where you'll still have to walk the entire tree? The children located at indexes that had children before will always have a valid layout offset, no?

Reply:
no you are right. We won't have that as long as I made the change suggested in https://github.com/flutter/flutter/pull/42703/files/9e0777c4e0be3c5f57260d58e9e0d8ad9b6c7229#r341749263